central/image/service/service_impl.go

```diff
@@ -248,18 +277,10 @@ func (s *serviceImpl) ScanImageInternal(ctx context.Context, request *v1.ScanIma
 	}
 
 	if _, err := s.enricher.EnrichImage(ctx, enrichmentContext, img); err != nil {
-		log.Errorf("error enriching image %q: %v", request.GetImage().GetName().GetFullName(), err)
-		// purposefully, don't return here because we still need to save it into the DB so there is a reference
-		// even if we weren't able to enrich it
-	}
-
-	// Due to discrepancies in digests retrieved from metadata pulls and k8s, only upsert if the request
-	// contained a digest
-	if imgID != "" {
-		_ = s.saveImage(img)
+		log.Errorf("error enriching image %q: %v", img.GetName().GetFullName(), err)
+		// Purposefully, don't return here because we still need to save it into the DB so there is a reference
```

Comment:
Are we sure saving the result of this `enricher.EnrichImage` when:
1. Error is returned
2. Image is already in the datastore but request image name isn't https://github.com/stackrox/stackrox/pull/3907/files#diff-1969ddcb957be51b7db622434fa6b73050a9b853fe85c781790b74c7fc55078cR235

is the correct approach?

Before this PR, this code path only ignored image when there didn't exist an image already in datastore. Now we can actually replace the existing image.

Reply:
I think that's a fair point and good catch.

I've re-written `ScanImageInternal` to skip upserting the image in the database when:
- the image exists previously.
- the error from enriching the image is `!= nil`.